### PR TITLE
Fix dynamic batch dim crash on OpenVINO backend

### PR DESCRIPTION
**Pull request by goyaladitya05**
## Description of the change
This PR addresses a fix for the temporary turnaround introduced in keras-team/keras/pull/22681.

Replaced tensor.shape[0] with ops.shape(tensor)[0] in non_max_suppression and target_gather so tensor allocation works when the batch dim is None during OpenVINO graph tracing. ops.shape() returns a graph node on compiled backends and a plain int on eager backends.

## Checklist

- [x] I have added all the necessary unit tests for my change. NA
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: keras_hub/src/utils/tensor_utils.py [hunk 1]
```diff
@@ -451,9 +451,9 @@ def gather_unbatched(labels, match_indices, mask, mask_val):
 
         def assign_when_rows_empty():
             if len(labels.shape) > 1:
-                mask_shape = [match_indices.shape[0], labels.shape[-1]]
+                mask_shape = [ops.shape(match_indices)[0], labels.shape[-1]]
```
[the remark this answers is not in the document]
Comment (goyaladitya05) [reply keras_hub/src/utils/tensor_utils.py:454]: ops.shape(labels)[-1] is the last dimension is the number of classes or box coordinates, a fixed constant which is always known.

**Review by hertschuh — APPROVED** [review]
Thanks. But how did you test this?

**Comment by goyaladitya05** [comment]
@hertschuh @sachinprasadhs this was my analysis of the issue. 

_get_compiled_model always calls _parameterize_data(data, dynamic_batch_size=True), setting ov_shape[0] = -1 for every input on every predict()/generate() call. This makes the batch dim dynamic on the OV parameter node.

ops.shape(x) correctly handles this for dynamic dims, it returns OpenVINOKerasTensor scalars (via shape_of + gather), not Python None. So _build_cache does the right thing:

```
batch_size = ops.shape(token_ids)[0]  # -> OpenVINOKerasTensor scalar 
ops.zeros([batch_size, num_layers, 2, max_len, heads, dim])  # -> crash
```

The crash is in zeros/ones/full/empty, which all call ov_opset.constant(shape, Type.i32) directly. ov_opset.constant cannot serialize an OpenVINOKerasTensor object, as it expects Python ints. This is the one of the reasons for shape_to_ov_output() implementation.

I believe 2 different fixes are required, the current PR being one of them.

1. [Keras PR #22747 (unmerged)](https://github.com/keras-team/keras/pull/22747): replace ov_opset.constant(shape) with shape_to_ov_output() in zeros, ones, full, and empty. This handles mixed tuples of plain ints and OpenVINOKerasTensor scalars by building per-dim constants and concatenating them into a valid i32 shape tensor.

2. this PR:  x.shape[0] returns Python None (the cached .shape attribute), bypassing the shape_of + gather logic in ops.shape(). Without this fix, even after #22747 lands, _build_cache would pass None into zeros and hit the new guard in shape_to_ov_output that explicitly rejects None with "use ops.shape(x) instead of x.shape".

Both are needed: #22747 without this PR,  x.shape[0] still passes None, hits the guard, crashes. This PR without #22747, ops.shape(x)[0] returns an OV scalar, but zeros still calls ov_opset.constant which rejects it.

please correct me if I'm wrong anywhere.

**Comment by hertschuh** [comment]
> @hertschuh @sachinprasadhs this was my analysis of the issue.
> 
> _get_compiled_model always calls _parameterize_data(data, dynamic_batch_size=True), setting ov_shape[0] = -1 for every input on every predict()/generate() call. This makes the batch dim dynamic on the OV parameter node.
> 
> ops.shape(x) correctly handles this for dynamic dims, it returns OpenVINOKerasTensor scalars (via shape_of + gather), not Python None. So _build_cache does the right thing:
> 
> ```
> batch_size = ops.shape(token_ids)[0]  # -> OpenVINOKerasTensor scalar 
> ops.zeros([batch_size, num_layers, 2, max_len, heads, dim])  # -> crash
> ```
> 
> The crash is in zeros/ones/full/empty, which all call ov_opset.constant(shape, Type.i32) directly. ov_opset.constant cannot serialize an OpenVINOKerasTensor object, as it expects Python ints. This is the one of the reasons for shape_to_ov_output() implementation.
> 
> I believe 2 different fixes are required, the current PR being one of them.
> 
> 1. [Keras PR #22747 (unmerged)](https://github.com/keras-team/keras/pull/22747): replace ov_opset.constant(shape) with shape_to_ov_output() in zeros, ones, full, and empty. This handles mixed tuples of plain ints and OpenVINOKerasTensor scalars by building per-dim constants and concatenating them into a valid i32 shape tensor.
> 2. this PR:  x.shape[0] returns Python None (the cached .shape attribute), bypassing the shape_of + gather logic in ops.shape(). Without this fix, even after #22747 lands, _build_cache would pass None into zeros and hit the new guard in shape_to_ov_output that explicitly rejects None with "use ops.shape(x) instead of x.shape".
> 
> Both are needed: #22747 without this PR, x.shape[0] still passes None, hits the guard, crashes. This PR without #22747, ops.shape(x)[0] returns an OV scalar, but zeros still calls ov_opset.constant which rejects it.
> 
> please correct me if I'm wrong anywhere.

Ok so https://github.com/keras-team/keras/pull/22747 is merged.

This is safe to merge I believe.

Then you'll have another PR to remove the `dynamic_batch_size` option in `_parameterize_data`, right?

**Comment by goyaladitya05** [comment]
> Ok so [keras-team/keras#22747](https://github.com/keras-team/keras/pull/22747) is merged.
> 
> This is safe to merge I believe.

Yes, but @sachinprasadhs, could you please share how initially it was tested that dynamic batch was being broken here? I would really prefer to validate that case properly, before merging.

> Then you'll have another PR to remove the `dynamic_batch_size` option in `_parameterize_data`, right?

Yeah, I'll open a PR on that.

**Comment by hertschuh** [comment]
> > Ok so [keras-team/keras#22747](https://github.com/keras-team/keras/pull/22747) is merged.
> > This is safe to merge I believe.
> 
> Yes, but @sachinprasadhs, could you please share how initially it was tested that dynamic batch size was being broken here? I would really prefer to validate that case properly, before merging.

- You clone both `keras` and `keras-hub` locally
- You install keras from the cloned source by going in the `keras` folder and running `pip install -e .`
- Then within the `keras-hub` folder you run: 

```shell
KERAS_BACKEND=openvino pytest keras_hub/src/models/gpt2/gpt2_causal_lm_test.py
```

**Comment by goyaladitya05** [comment]
> * You clone both `keras` and `keras-hub` locally
> * You install keras from the cloned source by going in the `keras` folder and running `pip install -e .`
> * Then within the `keras-hub` folder you run:
> 
> ```shell
> KERAS_BACKEND=openvino pytest keras_hub/src/models/gpt2/gpt2_causal_lm_test.py
> ```

Okay, I'll use these steps, verify everything works and report back here. In the meanwhile, please do not merge this PR for now.

**Comment by goyaladitya05** [comment]
The tests still fail, some more fixes must be required. I haven't investigated deeper as I'm a little bit occupied right now, hence putting this on hold for a while.

**Comment by goyaladitya05** [comment]
@hertschuh @sachinprasadhs 
I was partially wrong earlier. No changes are required in keras-hub. The following 2 fixes are required, I validated using the steps given earlier.

1. https://github.com/keras-team/keras/pull/22747

2. https://github.com/keras-team/keras/pull/22915
